sql/analyzer/resolve_subqueries.go

```diff
@@ -142,3 +159,20 @@ func cacheSubqueryResults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scop
 		return s, nil
 	})
 }
+
+// cacheSubqueryAlisesInJoins will look for joins against subquery aliases that
+// will repeatedly execute the subquery, and will insert a *plan.CachedResults
+// node on top of those nodes when it is safe to do so.
+func cacheSubqueryAlisesInJoins(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	return plan.TransformUpWithParent(n, func(child, parent sql.Node, childNum int) (sql.Node, error) {
+		_, isJoin := parent.(plan.JoinNode)
+		_, isIndexedJoin := parent.(*plan.IndexedJoin)
+		if isJoin || isIndexedJoin {
+			sa, isSubqueryAlias := child.(*plan.SubqueryAlias)
```

Comment:
This is going to waste memory unnecessarily if the subquery is the first table in the join and therefore unnecessary to cache. Maybe add a TODO to fix? (algorithm to determine this case doesn't map easily to one of the transform up methods, i don't think)